src/components/connectAdvanced.js

```diff
@@ -124,54 +105,126 @@ export default function connectAdvanced(
 
         this.version = version
         this.renderCount = 0
-        this.store = props[storeKey] || context[storeKey]
+        const store = props[storeKey] || context[storeKey]
         this.propsMode = Boolean(props[storeKey])
         this.setWrappedInstance = this.setWrappedInstance.bind(this)
 
-        invariant(this.store,
+        invariant(store,
           `Could not find "${storeKey}" in either the context or props of ` +
           `"${displayName}". Either wrap the root component in a <Provider>, ` +
           `or explicitly pass "${storeKey}" as a prop to "${displayName}".`
         )
-
+        const storeState = store.getState()
+        const subscription = (this.propsMode ? null : context[subscriptionKey])
+        const childPropsSelector = this.createChildSelector(store)
+        this.state = {
+          props,
+          childPropsSelector,
+          childProps: {},
+          store,
+          storeState,
+          error: null,
+          subscription: new Subscription(store, subscription, this.updateChildPropsFromReduxStore.bind(this)),
+          lastNotify: noop,
+          notifyNestedSubs: noop
+        }
         this.state = {
-          updater: this.createUpdater()
+          ...this.state,
+          ...Connect.getChildPropsState(props, this.state)
+        }
+      }
+
+      static getChildPropsState(props, state) {
+        try {
+          const nextProps = state.childPropsSelector(state.store.getState(), props)
```

Comment:
Reading the store state here is still my biggest concern, async-wise.  But, I guess we can go with it until we get more examples from the React team for what proper patterns look like.